parlai/agents/starspace_torch/starspace_torch.py

```diff
@@ -0,0 +1,367 @@
+    """
+
+    OPTIM_OPTS = {
+        'adadelta': optim.Adadelta,
+        'adagrad': optim.Adagrad,
+        'adam': optim.Adam,
+        'adamax': optim.Adamax,
+        'asgd': optim.ASGD,
+        'lbfgs': optim.LBFGS,
+        'rmsprop': optim.RMSprop,
+        'rprop': optim.Rprop,
+        'sgd': optim.SGD,
+    }
+
+    @staticmethod
+    def dictionary_class():
+        return DictionaryAgent
+
+    @staticmethod
+    def add_cmdline_args(argparser):
+        """Add command-line arguments specifically for this agent."""
+        TorchRankerAgent.add_cmdline_args(argparser)
+        agent = argparser.add_argument_group('StarSpace Arguments')
+        agent.add_argument(
+            '-esz', '--embeddingsize', type=int, default=128,
+            help='size of the token embeddings')
+        agent.add_argument(
+            '-enorm', '--embeddingnorm', type=float, default=10,
+            help='max norm of word embeddings')
+        agent.add_argument(
+            '-shareEmb', '--share-embeddings', type='bool', default=True,
+            help='whether LHS and RHS share embeddings')
+        agent.add_argument(
+            '--lins', default=0, type=int,
+            help='If set to 1, add a linear layer between lhs and rhs.')
+        agent.add_argument(
+            '-lr', '--learningrate', type=float, default=0.1,
+            help='learning rate')
+        agent.add_argument(
+            '-margin', '--margin', type=float, default=0.1,
+            help='margin')
+        agent.add_argument(
+            '--input-dropout', type=float, default=0,
+            help='fraction of input/output features to dropout during training')
+        agent.add_argument(
+            '-opt', '--optimizer', default='sgd',
+            choices=StarspaceTorchAgent.OPTIM_OPTS.keys(),
+            help='Choose between pytorch optimizers. '
+                 'Any member of torch.optim is valid and will '
+                 'be used with default params except learning '
+                 'rate (as specified by -lr).')
+        agent.add_argument(
+            '-tr', '--truncate', default=1000, type=int,
+            help='Truncate input lengths to increase speed / '
+                 'use less memory.')
+        agent.add_argument(
+            '-k', '--neg-samples', type=int, default=10,
+            help='number k of negative samples per example')
+        agent.add_argument(
+            '--parrot-neg', type=bool, default=False,
+            help='include query as a negative')
+        agent.add_argument(
+            '--tfidf', type='bool', default=False,
+            help='Use frequency based normalization for embeddings.')
+        agent.add_argument(
+            '-cs', '--cache-size', type=int, default=1000,
+            help='size of negative sample cache to draw from')
+        agent.add_argument(
+            '-cands', '--candidates', type=str, default='custom',
```

Comment:
These are provided by TRA, yes? No need to override them here.

Reply:
overridden because the default here is 'custom'